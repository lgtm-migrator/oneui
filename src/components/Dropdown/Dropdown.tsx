import React, { useState } from 'react';
import { useSelect, PropGetters } from 'downshift';
import { usePopper } from 'react-popper';
import { bem } from '../../utils/bem/bem';
import { mergeRefs } from '../../utils/mergeRefs';
import { PopupPlacement } from '../../constants';
import { ButtonProps } from '../Buttons';
import { List } from '../List';
import styles from './Dropdown.scss';

const { elem } = bem('Dropdown', styles);

export type DropdownRenderArgs<V> = {
    getItemPropsByIndex: (index: number) => PropGetters<V>;
};

interface Props<V> extends Omit<React.HTMLAttributes<HTMLDivElement>, 'onChange'> {
    /** List of items */
    items: V[];
    /** Popup content */
    children: (args: DropdownRenderArgs<V>) => React.ReactElement;
    /** {@link Button} element, controlled by current component */
    button: React.FunctionComponentElement<ButtonProps>;
    /** onChange called when an item is selected */
    onChange: (value: V) => void;
    /** onOpen called on request to be open */
    onOpen?: () => void;
    /** onClose called on request to be close */
    onClose?: () => void;
    /** Should the children be shown. This prop is used when we need to have Dropdown as "controlled" component */
    isOpen?: boolean;
    /**
     * Convert item to string.
     * Prop is required if passed items are not list of strings.
     * This is need for accessibility aria-live messages for downshift.
     */
    itemToString?: (item: V | null) => string;
    /** Popup placement relative to button */
    placement?: PopupPlacement;
}

/**
 * Component is intended for displaying available actions by button click.
 * It allows you to navigate with keyboard across actions.
 * If you don't need navigation - use PopupBase component.
 */
export function Dropdown<V>(props: Props<V>) {
    const {
        button,
        children,
        items,
        onChange,
        onOpen,
        onClose,
        isOpen: isOpenProp,
        itemToString,
        placement,
        ...rest
    } = props;

    // Assuming the rest of the items have the same type as the first item
    if (items[0] && typeof items[0] !== 'string' && itemToString === undefined) {
        throw new Error('You need pass "itemToString" for non string "items"');
    }

    const [referenceElement, setReferenceElement] = useState(null);
    const [popperElement, setPopperElement] = useState(null);

    const state = usePopper(referenceElement, popperElement, {
        placement,
        modifiers: [
            {
                name: 'offset',
                options: { offset: [0, 3] },
            },
        ],
    });

    const {
        isOpen,
        getItemProps,
        getMenuProps,
        getToggleButtonProps,
        highlightedIndex,
        reset,
    } = useSelect<V>({
        items,
        isOpen: isOpenProp,
        itemToString,
        onSelectedItemChange: ({ selectedItem }) => {
            if (selectedItem) {
                onChange(selectedItem);
                reset();
            }
        },
        onIsOpenChange: (changes) => {
            if (changes.isOpen) {
                if (onOpen) {
                    onOpen();
                }
            } else if (onClose) {
                onClose();
            }
        },
    });

    const menuProps = getMenuProps();
    const toggleButtonProps = getToggleButtonProps();
    const openPopperProps = isOpen && {
        style: state.styles.popper,
        ...state.attributes.popper,
        ...elem('list'),
    };
    const getItemPropsByIndex = (index: number) => {
        return {
            ...getItemProps({
                index,
                item: items[index],
            }),
            isHighlighted: highlightedIndex === index,
        };
    };

    return (
        <>
            {React.cloneElement(button, {
                ...rest,
                ...toggleButtonProps,
                ref: mergeRefs([setReferenceElement, toggleButtonProps.ref, button.ref]),
            })}
            <List
                {...menuProps}
                {...openPopperProps}
                ref={mergeRefs([isOpen && setPopperElement, menuProps.ref])}
                isControlledNavigation
            >
                {isOpen && children({ getItemPropsByIndex })}
            </List>
        </>
    );
}

Dropdown.displayName = 'Dropdown';

Dropdown.defaultProps = {
    placement: 'bottom-end',
    onOpen: undefined,
    onClose: undefined,
    isOpen: undefined,
    itemToString: undefined,
};
