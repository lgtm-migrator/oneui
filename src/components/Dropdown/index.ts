export { Dropdown, DropdownRenderArgs } from './Dropdown';
