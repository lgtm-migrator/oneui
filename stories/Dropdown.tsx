import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { select, withKnobs } from '@storybook/addon-knobs';
import { Button, Dropdown, ListItem, IconTextkernel } from '@textkernel/oneui';
import { POPUP_PLACEMENTS } from '@textkernel/oneui/constants';
import { HiDotsVertical } from 'react-icons/hi';

const styles = {
    content: {
        display: 'flex',
        justifyContent: 'center',
        padding: '0 40px',
    },
    customDiv: {
        alignItems: 'center',
        background: '#ffff003d',
        display: 'flex',
        height: '40px',
        justifyContent: 'center',
    },
    listItem: {
        alignItems: 'center',
        display: 'flex',
    },
    divider: {
        borderBottom: '1px solid #80808045',
    },
    icon: {
        marginRight: '10px',
    },
};

storiesOf('Molecules|Dropdown', module)
    .addDecorator(withKnobs)
    .add('Dropdown', () => {
        const onChange = (value) => {
            console.log(`onChange was called value : ${value}`);
        };
        const onOpen = () => {
            console.log('Dropdown was requested to be open.');
        };
        const onClose = () => {
            console.log('Dropdown was requested to be close.');
        };

        const customButtonsDemo = [
            <Button context="brand">Click me!</Button>,
            <Button context="neutral">
                <HiDotsVertical />
            </Button>,
            <Button context="link">Select any</Button>,
        ];

        const buttonIndex = select('customButton', [0, 1, 2], 0);

        const customValues = [
            'ListItem with value 1',
            'ListItem with value 2',
            'ListItem with value 3',
        ];

        return (
            <div style={styles.content}>
                <Dropdown<string>
                    items={customValues}
                    style={{ width: 'fit-content' }}
                    button={customButtonsDemo[buttonIndex]}
                    onChange={onChange}
                    onOpen={onOpen}
                    onClose={onClose}
                    placement={select('placement', POPUP_PLACEMENTS, 'bottom-end')}
                >
                    {({ getItemPropsByIndex }) => {
                        return (
                            <>
                                <ListItem disabled>Disabled</ListItem>
                                {customValues.map((value, index) => (
                                    <ListItem key={value} {...getItemPropsByIndex(index)}>
                                        <div style={styles.listItem}>
                                            <IconTextkernel context="brand" style={styles.icon} />
                                            {value}
                                        </div>
                                    </ListItem>
                                ))}
                                <div style={styles.customDiv}>Just custom div element</div>
                            </>
                        );
                    }}
                </Dropdown>
            </div>
        );
    });
