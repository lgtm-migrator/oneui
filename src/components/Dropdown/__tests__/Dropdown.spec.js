import React from 'react';
import toJson from 'enzyme-to-json';
import { Dropdown } from '../Dropdown';
import { Button } from '../../Buttons';
import { ListItem } from '../../List';

describe('Dropdown', () => {
    const mockOnChange = jest.fn();
    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <Dropdown
                items={['first-value']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <>
                        <ListItem disabled>Disabled</ListItem>
                        <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                    </>
                )}
            </Dropdown>
        );
    });

    it('should render correctly closed', () => {
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('ListItem')).toHaveLength(0);
        wrapper.unmount();
    });

    it('should render correctly opened', () => {
        wrapper.find('button').simulate('click');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('List')).toHaveLength(1);
        expect(wrapper.find('ListItem')).toHaveLength(2);
        wrapper.unmount();
    });

    it('should downshift only by enabled items with value', () => {
        wrapper = mount(
            <Dropdown
                items={['1', '2']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <>
                        <ListItem disabled>Disabled</ListItem>
                        <ListItem {...getItemPropsByIndex(0)}>With value 1</ListItem>
                        <ListItem>Without value</ListItem>
                        <div className="customDiv">Div</div>
                        <ListItem {...getItemPropsByIndex(1)}>With value 2</ListItem>
                    </>
                )}
            </Dropdown>
        );
        wrapper.find('button').simulate('click');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('ul').children()).toHaveLength(5);

        const keyDown = () => wrapper.find('List').simulate('keyDown', { key: 'ArrowDown' });
        const findHighlighted = () =>
            wrapper
                .find('ul')
                .children()
                .filterWhere((component) => component.props().isHighlighted === true);

        // 1 keydown
        keyDown();
        const highlightedItem1 = findHighlighted();
        expect(highlightedItem1).toHaveLength(1);
        expect(highlightedItem1.text()).toContain('With value 1');

        // 2 keydown
        keyDown();
        const highlightedItem2 = findHighlighted();
        expect(highlightedItem2).toHaveLength(1);
        expect(highlightedItem2.text()).toContain('With value 2');

        // 3 keydown. Should be again last not disabled with value item => 'With value 2'
        keyDown();
        const highlightedItem3 = findHighlighted();
        expect(highlightedItem3).toHaveLength(1);
        expect(highlightedItem3.text()).toContain('With value 2');

        wrapper.unmount();
    });

    it('onChange should return passed value', () => {
        wrapper = mount(
            <Dropdown
                items={['testValue']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <>
                        <ListItem disabled>Disabled</ListItem>
                        <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                    </>
                )}
            </Dropdown>
        );
        wrapper.find('button').simulate('click');

        wrapper.find('List').simulate('keyDown', { key: 'ArrowDown' });
        wrapper.find('List').simulate('keyDown', { key: 'Enter' });

        expect(mockOnChange).toBeCalledWith('testValue');

        wrapper.unmount();
    });

    it('should render correctly with mixed children: array and single ListItem', () => {
        wrapper = mount(
            <Dropdown
                items={['one', 'two']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <>
                        <ListItem disabled>Disabled</ListItem>
                        {['one', 'two'].map((value, index) => (
                            <ListItem key={value} {...getItemPropsByIndex(index)}>
                                {value}
                            </ListItem>
                        ))}
                    </>
                )}
            </Dropdown>
        );
        wrapper.find('button').simulate('click');
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('ListItem')).toHaveLength(3);
        wrapper.unmount();
    });

    it('isOpen should open and close list of items', () => {
        wrapper = mount(
            <Dropdown
                isOpen
                items={['testValue']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                )}
            </Dropdown>
        );
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('ListItem')).toHaveLength(1);
        wrapper.setProps({ isOpen: false });
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(wrapper.find('ListItem')).toHaveLength(0);
        wrapper.unmount();
    });

    it('onOpen and onClose should fired on button click', () => {
        const mockOnOpen = jest.fn();
        const mockOnClose = jest.fn();
        wrapper = mount(
            <Dropdown
                items={['testValue']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                onOpen={mockOnOpen}
                onClose={mockOnClose}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                )}
            </Dropdown>
        );
        wrapper.find('button').simulate('click');

        expect(mockOnOpen).toBeCalledTimes(1);
        expect(mockOnClose).toBeCalledTimes(0);

        wrapper.find('button').simulate('click');

        expect(mockOnOpen).toBeCalledTimes(1);
        expect(mockOnClose).toBeCalledTimes(1);

        wrapper.unmount();
    });

    it('onClose should fired after selection item', () => {
        const mockOnClose = jest.fn();
        wrapper = mount(
            <Dropdown
                items={['testValue']}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                onClose={mockOnClose}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                )}
            </Dropdown>
        );
        wrapper.find('button').simulate('click');

        wrapper.find('List').simulate('keyDown', { key: 'ArrowDown' });
        wrapper.find('List').simulate('keyDown', { key: 'Enter' });

        expect(mockOnClose).toBeCalledTimes(1);

        wrapper.unmount();
    });

    it('should throw error while itemToString is missed and items are not strings', () => {
        expect(() =>
            shallow(
                <Dropdown
                    items={[{ value: 'testValue' }]}
                    button={<Button context="brand">Click me!</Button>}
                    onChange={mockOnChange}
                    placement="top-start"
                >
                    {({ getItemPropsByIndex }) => (
                        <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                    )}
                </Dropdown>
            )
        ).toThrow('You need pass "itemToString" for non string "items"');
    });

    it('itemToString should correctly convert', () => {
        wrapper = mount(
            <Dropdown
                items={[{ value: 'testValue' }]}
                button={<Button context="brand">Click me!</Button>}
                onChange={mockOnChange}
                itemToString={(item) => item.value}
                placement="top-start"
            >
                {({ getItemPropsByIndex }) => (
                    <>
                        <ListItem disabled>Disabled</ListItem>
                        <ListItem {...getItemPropsByIndex(0)}>With value</ListItem>
                    </>
                )}
            </Dropdown>
        );
        wrapper.find('button').simulate('click');

        wrapper.find('List').simulate('keyDown', { key: 'ArrowDown' });
        wrapper.find('List').simulate('keyDown', { key: 'Enter' });

        expect(mockOnChange).toBeCalledWith({ value: 'testValue' });

        wrapper.unmount();
    });
});
