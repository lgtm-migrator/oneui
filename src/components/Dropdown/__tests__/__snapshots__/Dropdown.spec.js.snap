// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown isOpen should open and close list of items 1`] = `
<Dropdown
  button={
    <ForwardRef
      context="brand"
      disabled={false}
      isBlock={false}
      isInline={false}
      size="normal"
      type="button"
    >
      Click me!
    </ForwardRef>
  }
  isOpen={true}
  items={
    Array [
      "testValue",
    ]
  }
  onChange={[MockFunction]}
  placement="top-start"
>
  <Button
    aria-expanded={true}
    aria-haspopup="listbox"
    aria-labelledby="downshift-50-label downshift-50-toggle-button"
    context="brand"
    disabled={false}
    id="downshift-50-toggle-button"
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    size="normal"
    type="button"
  >
    <button
      aria-expanded={true}
      aria-haspopup="listbox"
      aria-labelledby="downshift-50-label downshift-50-toggle-button"
      className="Button Button--context_brand"
      disabled={false}
      id="downshift-50-toggle-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me!
    </button>
  </Button>
  <List
    aria-labelledby="downshift-50-label"
    className="Dropdown__list"
    doSelectOnNavigate={false}
    id="downshift-50-menu"
    isControlledNavigation={true}
    isDivided={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
    role="listbox"
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
    tabIndex={-1}
  >
    <ul
      aria-labelledby="downshift-50-label"
      className="List Dropdown__list"
      id="downshift-50-menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="listbox"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
      tabIndex={-1}
    >
      <ListItem
        aria-selected="false"
        className="List__item"
        disabled={false}
        highlightContext="default"
        id="downshift-50-item-0"
        isHighlighted={false}
        isSelected={false}
        key=".0"
        onClick={[Function]}
        onMouseMove={[Function]}
        role="option"
      >
        <li
          aria-selected="false"
          className="ListItem ListItem--clickable List__item"
          id="downshift-50-item-0"
          onMouseMove={[Function]}
          role="option"
        >
          <div
            className="ListItem__container"
            onClick={[Function]}
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                With value
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
    </ul>
  </List>
</Dropdown>
`;

exports[`Dropdown isOpen should open and close list of items 2`] = `
<Dropdown
  button={
    <ForwardRef
      context="brand"
      disabled={false}
      isBlock={false}
      isInline={false}
      size="normal"
      type="button"
    >
      Click me!
    </ForwardRef>
  }
  isOpen={false}
  items={
    Array [
      "testValue",
    ]
  }
  onChange={[MockFunction]}
  placement="top-start"
>
  <Button
    aria-expanded={false}
    aria-haspopup="listbox"
    aria-labelledby="downshift-50-label downshift-50-toggle-button"
    context="brand"
    disabled={false}
    id="downshift-50-toggle-button"
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    size="normal"
    type="button"
  >
    <button
      aria-expanded={false}
      aria-haspopup="listbox"
      aria-labelledby="downshift-50-label downshift-50-toggle-button"
      className="Button Button--context_brand"
      disabled={false}
      id="downshift-50-toggle-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me!
    </button>
  </Button>
  <List
    aria-labelledby="downshift-50-label"
    doSelectOnNavigate={false}
    id="downshift-50-menu"
    isControlledNavigation={true}
    isDivided={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
    role="listbox"
    tabIndex={-1}
  >
    <ul
      aria-labelledby="downshift-50-label"
      className="List"
      id="downshift-50-menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="listbox"
      tabIndex={-1}
    />
  </List>
</Dropdown>
`;

exports[`Dropdown should downshift only by enabled items with value 1`] = `
<Dropdown
  button={
    <ForwardRef
      context="brand"
      disabled={false}
      isBlock={false}
      isInline={false}
      size="normal"
      type="button"
    >
      Click me!
    </ForwardRef>
  }
  items={
    Array [
      "1",
      "2",
    ]
  }
  onChange={[MockFunction]}
  placement="top-start"
>
  <Button
    aria-expanded={true}
    aria-haspopup="listbox"
    aria-labelledby="downshift-12-label downshift-12-toggle-button"
    context="brand"
    disabled={false}
    id="downshift-12-toggle-button"
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    size="normal"
    type="button"
  >
    <button
      aria-expanded={true}
      aria-haspopup="listbox"
      aria-labelledby="downshift-12-label downshift-12-toggle-button"
      className="Button Button--context_brand"
      disabled={false}
      id="downshift-12-toggle-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me!
    </button>
  </Button>
  <List
    aria-labelledby="downshift-12-label"
    className="Dropdown__list"
    doSelectOnNavigate={false}
    id="downshift-12-menu"
    isControlledNavigation={true}
    isDivided={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
    role="listbox"
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
    tabIndex={-1}
  >
    <ul
      aria-labelledby="downshift-12-label"
      className="List Dropdown__list"
      id="downshift-12-menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="listbox"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
      tabIndex={-1}
    >
      <ListItem
        disabled={true}
        highlightContext="default"
        isHighlighted={false}
        isSelected={false}
      >
        <li
          className="ListItem ListItem--disabled"
        >
          <div
            className="ListItem__container"
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                Disabled
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
      <ListItem
        aria-selected="false"
        disabled={false}
        highlightContext="default"
        id="downshift-12-item-0"
        isHighlighted={false}
        isSelected={false}
        onClick={[Function]}
        onMouseMove={[Function]}
        role="option"
      >
        <li
          aria-selected="false"
          className="ListItem ListItem--clickable"
          id="downshift-12-item-0"
          onMouseMove={[Function]}
          role="option"
        >
          <div
            className="ListItem__container"
            onClick={[Function]}
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                With value 1
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
      <ListItem
        disabled={false}
        highlightContext="default"
        isHighlighted={false}
        isSelected={false}
      >
        <li
          className="ListItem"
        >
          <div
            className="ListItem__container"
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                Without value
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
      <div
        className="customDiv"
      >
        Div
      </div>
      <ListItem
        aria-selected="false"
        disabled={false}
        highlightContext="default"
        id="downshift-12-item-1"
        isHighlighted={false}
        isSelected={false}
        onClick={[Function]}
        onMouseMove={[Function]}
        role="option"
      >
        <li
          aria-selected="false"
          className="ListItem ListItem--clickable"
          id="downshift-12-item-1"
          onMouseMove={[Function]}
          role="option"
        >
          <div
            className="ListItem__container"
            onClick={[Function]}
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                With value 2
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
    </ul>
  </List>
</Dropdown>
`;

exports[`Dropdown should render correctly closed 1`] = `
<Dropdown
  button={
    <ForwardRef
      context="brand"
      disabled={false}
      isBlock={false}
      isInline={false}
      size="normal"
      type="button"
    >
      Click me!
    </ForwardRef>
  }
  items={
    Array [
      "first-value",
    ]
  }
  onChange={[MockFunction]}
  placement="top-start"
>
  <Button
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label downshift-0-toggle-button"
    context="brand"
    disabled={false}
    id="downshift-0-toggle-button"
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    size="normal"
    type="button"
  >
    <button
      aria-haspopup="listbox"
      aria-labelledby="downshift-0-label downshift-0-toggle-button"
      className="Button Button--context_brand"
      disabled={false}
      id="downshift-0-toggle-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me!
    </button>
  </Button>
  <List
    aria-labelledby="downshift-0-label"
    doSelectOnNavigate={false}
    id="downshift-0-menu"
    isControlledNavigation={true}
    isDivided={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
    role="listbox"
    tabIndex={-1}
  >
    <ul
      aria-labelledby="downshift-0-label"
      className="List"
      id="downshift-0-menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="listbox"
      tabIndex={-1}
    />
  </List>
</Dropdown>
`;

exports[`Dropdown should render correctly opened 1`] = `
<Dropdown
  button={
    <ForwardRef
      context="brand"
      disabled={false}
      isBlock={false}
      isInline={false}
      size="normal"
      type="button"
    >
      Click me!
    </ForwardRef>
  }
  items={
    Array [
      "first-value",
    ]
  }
  onChange={[MockFunction]}
  placement="top-start"
>
  <Button
    aria-expanded={true}
    aria-haspopup="listbox"
    aria-labelledby="downshift-3-label downshift-3-toggle-button"
    context="brand"
    disabled={false}
    id="downshift-3-toggle-button"
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    size="normal"
    type="button"
  >
    <button
      aria-expanded={true}
      aria-haspopup="listbox"
      aria-labelledby="downshift-3-label downshift-3-toggle-button"
      className="Button Button--context_brand"
      disabled={false}
      id="downshift-3-toggle-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me!
    </button>
  </Button>
  <List
    aria-labelledby="downshift-3-label"
    className="Dropdown__list"
    doSelectOnNavigate={false}
    id="downshift-3-menu"
    isControlledNavigation={true}
    isDivided={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
    role="listbox"
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
    tabIndex={-1}
  >
    <ul
      aria-labelledby="downshift-3-label"
      className="List Dropdown__list"
      id="downshift-3-menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="listbox"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
      tabIndex={-1}
    >
      <ListItem
        disabled={true}
        highlightContext="default"
        isHighlighted={false}
        isSelected={false}
      >
        <li
          className="ListItem ListItem--disabled"
        >
          <div
            className="ListItem__container"
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                Disabled
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
      <ListItem
        aria-selected="false"
        disabled={false}
        highlightContext="default"
        id="downshift-3-item-0"
        isHighlighted={false}
        isSelected={false}
        onClick={[Function]}
        onMouseMove={[Function]}
        role="option"
      >
        <li
          aria-selected="false"
          className="ListItem ListItem--clickable"
          id="downshift-3-item-0"
          onMouseMove={[Function]}
          role="option"
        >
          <div
            className="ListItem__container"
            onClick={[Function]}
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                With value
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
    </ul>
  </List>
</Dropdown>
`;

exports[`Dropdown should render correctly with mixed children: array and single ListItem 1`] = `
<Dropdown
  button={
    <ForwardRef
      context="brand"
      disabled={false}
      isBlock={false}
      isInline={false}
      size="normal"
      type="button"
    >
      Click me!
    </ForwardRef>
  }
  items={
    Array [
      "one",
      "two",
    ]
  }
  onChange={[MockFunction]}
  placement="top-start"
>
  <Button
    aria-expanded={true}
    aria-haspopup="listbox"
    aria-labelledby="downshift-41-label downshift-41-toggle-button"
    context="brand"
    disabled={false}
    id="downshift-41-toggle-button"
    isBlock={false}
    isInline={false}
    onClick={[Function]}
    onKeyDown={[Function]}
    size="normal"
    type="button"
  >
    <button
      aria-expanded={true}
      aria-haspopup="listbox"
      aria-labelledby="downshift-41-label downshift-41-toggle-button"
      className="Button Button--context_brand"
      disabled={false}
      id="downshift-41-toggle-button"
      onClick={[Function]}
      onKeyDown={[Function]}
      type="button"
    >
      Click me!
    </button>
  </Button>
  <List
    aria-labelledby="downshift-41-label"
    className="Dropdown__list"
    doSelectOnNavigate={false}
    id="downshift-41-menu"
    isControlledNavigation={true}
    isDivided={false}
    onBlur={[Function]}
    onKeyDown={[Function]}
    onMouseLeave={[Function]}
    role="listbox"
    style={
      Object {
        "left": "0",
        "position": "absolute",
        "top": "0",
      }
    }
    tabIndex={-1}
  >
    <ul
      aria-labelledby="downshift-41-label"
      className="List Dropdown__list"
      id="downshift-41-menu"
      onBlur={[Function]}
      onKeyDown={[Function]}
      onMouseLeave={[Function]}
      role="listbox"
      style={
        Object {
          "left": "0",
          "position": "absolute",
          "top": "0",
        }
      }
      tabIndex={-1}
    >
      <ListItem
        disabled={true}
        highlightContext="default"
        isHighlighted={false}
        isSelected={false}
      >
        <li
          className="ListItem ListItem--disabled"
        >
          <div
            className="ListItem__container"
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                Disabled
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
      <ListItem
        aria-selected="false"
        disabled={false}
        highlightContext="default"
        id="downshift-41-item-0"
        isHighlighted={false}
        isSelected={false}
        key="one"
        onClick={[Function]}
        onMouseMove={[Function]}
        role="option"
      >
        <li
          aria-selected="false"
          className="ListItem ListItem--clickable"
          id="downshift-41-item-0"
          onMouseMove={[Function]}
          role="option"
        >
          <div
            className="ListItem__container"
            onClick={[Function]}
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                one
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
      <ListItem
        aria-selected="false"
        disabled={false}
        highlightContext="default"
        id="downshift-41-item-1"
        isHighlighted={false}
        isSelected={false}
        key="two"
        onClick={[Function]}
        onMouseMove={[Function]}
        role="option"
      >
        <li
          aria-selected="false"
          className="ListItem ListItem--clickable"
          id="downshift-41-item-1"
          onMouseMove={[Function]}
          role="option"
        >
          <div
            className="ListItem__container"
            onClick={[Function]}
            role="presentation"
          >
            <Text
              context="default"
              inline={true}
              key=".0"
              size="normal"
            >
              <span>
                two
              </span>
            </Text>
          </div>
        </li>
      </ListItem>
    </ul>
  </List>
</Dropdown>
`;
